distributed/scheduler.py

```diff
@@ -5407,16 +5407,34 @@ def add_plugin(self, plugin=None, idempotent=False, **kwargs):
         if isinstance(plugin, type):
             plugin = plugin(self, **kwargs)
 
+        if name is None:
+            name = _get_plugin_name(plugin)
+
+        if self.has_plugin(name):
+            return
+
         if idempotent and any(isinstance(p, type(plugin)) for p in self.plugins):
             return
 
-        self.plugins.append(plugin)
+        self.plugins[name] = plugin
 
-    def remove_plugin(self, plugin):
+    def remove_plugin(self, plugin=None, name=None):
         """Remove external plugin from scheduler"""
-        self.plugins.remove(plugin)
+        if plugin is None and name is not None:
+            self.plugins.pop(name)
+        elif hasattr(plugin, "name"):
+            self.plugins.pop(plugin.name)
+        else:
+            raise ValueError(
+                "Plugins removed by instance must have a name attribute, "
+                "otherwise removal must be by name argument."
+            )
```

Comment:
There will be existing user code that looks like `scheduler.remove_plugin(plugin)` where `plugin` doesn't have a `name` attribute. Right now the changes in this PR will break such user code. We should try to be backwards compatible where it's reasonable to do so and raise a warning telling users to start using named scheduler plugins. Perhaps something along the lines of (note I've not tested any of the code below): 

```diff
diff --git a/distributed/scheduler.py b/distributed/scheduler.py
index 3c891ff3..6488dcac 100644
--- a/distributed/scheduler.py
+++ b/distributed/scheduler.py
@@ -5420,15 +5420,30 @@ class Scheduler(SchedulerState, ServerNode):

     def remove_plugin(self, plugin=None, name=None):
         """Remove external plugin from scheduler"""
-        if plugin is None and name is not None:
+        if plugin is not None:
+            warnings.warn(
+                "Removing scheduler plugin by value is deprecated and will be disabled in a future release. "
+                "Please remove scheduler plugins by name instead.",
+                category=FutureWarning,
+            )
+
+        if name is not None:
             self.plugins.pop(name)
         elif hasattr(plugin, "name"):
             self.plugins.pop(plugin.name)
         else:
-            raise ValueError(
-                "Plugins removed by instance must have a name attribute, "
-                "otherwise removal must be by name argument."
-            )
+            # TODO: Remove this block of code once removing plugins by value is disabled
+            if plugin in self.plugins.values():
+                if len(plugin is p for p in self.plugins.values()) > 1:
+                    # Use a better message here
+                    raise ValueError("Multiple plugins found")
+                else:
+                    warnings.warn(
+                        "Removing scheduler plugins by value is deprecated and will be disabled "
+                        "in a future release. Please add and remove scheduler plugins by name instead.",
+                        category=FutureWarning,
+                    )
+                    self.plugins = {n: p for n, p in self.plugins.items() if p is not plugin}

     def has_plugin(self, name):
         """Check if plugin exists by name."""
```

Reply:
Ah yea, got it. Adjusted this to take that into account.